tools/splitblocks/main.go

```diff
@@ -0,0 +1,232 @@
+			return nil
+		}
+
+		if err := splitBlock(ctx, cfg, bkt, blockMeta, logger); err != nil {
+			level.Error(logger).Log("msg", "failed to split block", "err", err)
+			return err
+		}
+
+		level.Info(logger).Log("msg", "block split successfully")
+		return nil
+	})
+}
+
+func listBlocks(ctx context.Context, bkt objstore.Bucket) ([]string, error) {
+	var blocks []string
+	err := bkt.Iter(ctx, "", func(name string) error {
+		if block, ok := block.IsBlockDir(name); ok {
+			blocks = append(blocks, block.String())
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return blocks, nil
+}
+
+// splitBlock downloads the source block to the output directory, then generates new blocks that are within cfg.blockRanges durations.
+// After all the splits succeed the original source block is removed from the output directory.
+func splitBlock(ctx context.Context, cfg config, bkt objstore.Bucket, meta block.Meta, logger log.Logger) error {
+	originalBlockDir := filepath.Join(cfg.outputDir, meta.ULID.String())
+	if err := block.Download(ctx, logger, bkt, meta.ULID, originalBlockDir); err != nil {
+		return err
+	}
+
+	_, err := splitLocalBlock(ctx, cfg.outputDir, originalBlockDir, meta, cfg.maxBlockDuration, logger)
+	return err
+}
+
+func splitLocalBlock(ctx context.Context, parentDir, blockDir string, meta block.Meta, maxDuration time.Duration, logger log.Logger) ([]ulid.ULID, error) {
+	origBlockMaxTime := meta.MaxTime
+	minTime := meta.MinTime
+	result := []ulid.ULID(nil)
+	for minTime < origBlockMaxTime {
+		// Max time cannot cross maxDuration boundary, but also should not be greater than original max time.
+		maxTime := min(timestamp.Time(minTime).Truncate(maxDuration).Add(maxDuration).UnixMilli(), origBlockMaxTime)
+
+		level.Info(logger).Log("msg", "splitting block", "minTime", timestamp.Time(minTime), "maxTime", timestamp.Time(maxTime))
+
+		// Inject a modified meta and abuse the repair into removing the now "outside" chunks.
+		// Chunks that cross boundaries are included in multiple blocks.
+		meta.MinTime = minTime
+		meta.MaxTime = maxTime
+		if err := meta.WriteToDir(logger, blockDir); err != nil {
+			return nil, errors.Wrap(err, "failed injecting meta for split")
+		}
+		splitID, err := block.Repair(ctx, logger, parentDir, meta.ULID, block.SplitBlocksSource, block.IgnoreCompleteOutsideChunk, block.IgnoreIssue347OutsideChunk)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed while splitting block")
+		}
+
+		splitMeta, err := block.ReadMetaFromDir(path.Join(parentDir, splitID.String()))
+		if err != nil {
+			return nil, errors.Wrap(err, "failed while reading meta.json from split block")
+		}
+
+		level.Info(logger).Log("msg", "created block from split", "minTime", timestamp.Time(minTime), "maxTime", timestamp.Time(maxTime), "splitID", splitID, "series", splitMeta.Stats.NumSeries, "chunks", splitMeta.Stats.NumChunks, "samples", splitMeta.Stats.NumSamples)
```

Comment:
nit: let's delete output blocks with 0 series.

Reply:
This has now been added.